### PR TITLE
Move more metadata and build flags from setup.py to pyproject.toml

### PR DESCRIPTION
**Pull request by hoechenberger**
Make better use of `pyproject.toml` by moving project metadata and build system settings there from `setup.py`. Now, basically all that `setup.py` does is dynamically generating package dependencies.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,11 +1,90 @@
...
 [tool.codespell]
 ignore-words = "ignore_words.txt"
 builtin = "clear,rare,informal,names,usage"
 skip = "doc/references.bib"
 
 [tool.ruff]
-select = ["E", "F", "W", "D"]
-exclude = ["__init__.py", "constants.py", "resources.py"]
+select = ["E/F/W/D"]
+exclude = ["__init__.py/constants.py/resources.py"]
```
Comment (hoechenberger) [inline pyproject.toml:87]: My TOML linter made these changes … does that make sense?? 🤔

**Comment by hoechenberger** [comment]
I don't know why the pre-commit bot keeps intervening … I rebased on `upstream/main` before pushing to this branch…

**Comment by hoechenberger** [comment]
@larsoner Any thoughts on [0096577](https://github.com/mne-tools/mne-python/pull/12055/commits/009657731f8f0c9c6b1d0f6a3ac2fd8369a3e0c3)?

**Comment by larsoner** [comment]
I don't know, but it seems like creating this branch from the `mne-tools/mne-python` rather than `hoechenberger/mne-python` is causing problems with CircleCI. Can you push a branch to your fork (maybe just your first commit rebased on `upstream/main`) and open a standard PR? If not I can do it quickly

**Comment by hoechenberger** [comment]
Ouch I'm so sorry, I never intended to push the branch here. Thanks for catching this, @larsoner!

**Comment by hoechenberger** [comment]
New PR #12058

**Pull request closed without merging**